packages/vscode-extension/src/hoverProvider.ts

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import * as vscode from "vscode";
+import { getPropertyByPath } from "@microsoft/teamsfx-core";
+import { manifestConfigDataRegex, manifestStateDataRegex } from "./constants";
+import { core, getSystemInputs } from "./handlers";
+
+export class ManifestTemplateHoverProvider implements vscode.HoverProvider {
+  public async provideHover(
+    document: vscode.TextDocument,
+    position: vscode.Position,
+    token: vscode.CancellationToken
+  ): Promise<vscode.Hover | undefined> {
+    const line = document.lineAt(position.line);
+
+    let regex;
+    let matches = manifestStateDataRegex.exec(line.text);
+    if (matches !== null) {
+      regex = manifestStateDataRegex;
+    } else {
+      matches = manifestConfigDataRegex.exec(line.text);
+      if (matches !== null) {
+        regex = manifestConfigDataRegex;
+      }
+    }
+
+    if (matches !== null && regex !== undefined) {
+      const key = matches[0].replace(/{/g, "").replace(/}/g, "");
+      const indexOf = line.text.indexOf(matches[0]);
+      const position = new vscode.Position(line.lineNumber, indexOf);
+      const range = document.getWordRangeAtPosition(
+        new vscode.Position(position.line, indexOf),
+        new RegExp(regex)
+      );
+      const message = await this.generateHoverMessage(key);
+      const hover = new vscode.Hover(message, range);
+      return hover;
+    }
+
+    return undefined;
+  }
+
+  private async generateHoverMessage(key: string): Promise<vscode.MarkdownString> {
+    const inputs = getSystemInputs();
+    const getConfigRes = await core.getProjectConfigV3(inputs);
+    if (getConfigRes.isErr()) throw getConfigRes.error;
+    const projectConfigs = getConfigRes.value;
+
+    let message = "";
+    if (projectConfigs && projectConfigs.envInfos) {
```

Comment:
If provision hasn't been done for one environment, will it show undefined and error pops up when user click view state file?

Reply:
yes, exactly. When user click view state file, and the environment has not been provisioned, it will show error like before.